text/1028-cli-hooks.rst

```diff
@@ -0,0 +1,159 @@
+we can introduce hooks for some of the schema altering commands.
+
+
+Motivation
+==========
+
+There are a number of tools that generate code based on the current schema
+(such as query-building tools, or query reflection tools). They need to be
+re-run after every schema change in order to keep the generated code in sync
+with the database. In larger projects the number of things that may need
+updating after a schema change is likely to grow. Migration hooks could make
+it easier to automate these tasks.
+
+There is also a longstanding need to implement some kind of mechanism for
+creating data fixtures. Unlike the migration hooks, fixtures are more likely
+to be run at project initialization or after a branch wipe command.
+
+
+Hooks Configuration
+===================
+
+All hooks are intended to be used within an active project, not for arbitrary
+CLI commands. Thus if a CLI command receives an explicit ``-I <instance>``
+option, the command should not trigger any project hooks even if the instance
+is linked to a project. Only commands that omit the instance and use the
+current project to determine the target instance are going to trigger hooks.
+
+The hooks will be described in the ``gel.toml`` file in the
+``[project-hooks]`` table. We define ``migration.apply.after``,
+``project.init.after``, ``branch.wipe.after``, and ``branch.switch.after``
+hook keys. The values for them are arrays of strings, where each string is
+going to be executed as a shell command. The naming scheme is intended to
+mimic the command structure to clearly indicate which commands will trigger
+the hooks. Current RFC only introduces hooks to be executed after a given
+command, but the naming scheme supports future "before" hooks as well.
+
+The scripts are executed in the order they appear in the array, sequentially.
+So if you have multiple steps that needs to be executed in a certain order,
+you can put the hook commands in the same order you would have executed them
+in the shell.
+
+All hooks will use the project root directory as the execution directory.
+
+
+Configuration Hooks
+-------------------
+
+This category of hooks are intended for one-time setup of important fixtures
+or application configuration. They are perfect for setting up data fixtures or
+exporting connection strings for non-gel tools to use. There are generally two
+scenarios when this kind of hook might be needed:
+
+1) After ``gel project init`` the ``project.init.after`` hook will be
+   executed. This is good for updating any additional configuration that
+   depends on the project database. It is also a good place for data fixture
+   scripts.
+
+2) After ``gel branch wipe`` the ``branch.wipe.after`` hook will be executed.
+   This may be a good place for restoring the data fixtures. However it is
+   probably unnecessary to run any configuration scripts at this time. This is
+   why this hook does not simply run all the same scripts as the
+   ``project.init.after``.
```

Comment:
@vpetrovykh Can you confirm that we also want `branch.wipe.after` to trigger after a `gel database wipe`?